File: 1528-kids-with-the-greatest-number-of-candies/kids-with-the-greatest-number-of-candies.py
# Brute Force Code & Optimal Code
class Solution:
    # return whether each child can have the greatest number of candies after receiving all extra candies
    def kidsWithCandies(self, candies: List[int], extraCandies: int) -> List[bool]:
        # find the maximum number of candies currently held by any child
        max_candies = max(candies)
        # store the result for each child
        result = []
        # check every child candy count
        for candy in candies:
            # calculte how many candies this child would have afterr receiving all extra candies
            new_candy_count = candy + extraCandies
            # check whether the child would have at least as many candies as the child whos currently has the most
            if new_candy_count >= max_candies:
                result.append(True)
            else:
                result.append(False)
        # return the result for all children
        return result 

# Time Complexity : O(N)
# Space Complexity : O(1)
        

File: 1528-kids-with-the-greatest-number-of-candies/kids-with-the-greatest-number-of-candies.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    // return whether each child can have the greatest number of candies after receiving all extra candies.
    vector<bool> kidsWithCandies(vector<int>& candies, int extraCandies) {
        // find the maximum number of candies currently held by any child.
        int maxCandies = *max_element(candies.begin(), candies.end());
        // store the result for each child.
        vector<bool> result;
        // check every child's candy count.
        for (int candy : candies) {
            // calculate how many candies this child would have after receiving all extra candies.
            int newCandyCount = candy + extraCandies;
            // check whether this child would have at least as many candies as the child who currently has the most candies.
            if (newCandyCount >= maxCandies) {
                result.push_back(true);
            }
            else {
                result.push_back(false);
            }
        }
        // return the result for all children.
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(1)